doc: add migration guide from nrf5 (#334)
nrfconnect/sdk-nrf-bm: doc/nrf-bm/migration/nrf5_bm_migration.rst

.. _nrf5_bm_migration:

> Bare metal migration guide

Reviewer: ```suggestion
Bare Metal migration guide
```

The build system and the |nRFVSC| extension both provide a way to conveniently browse and search all available project options and inspect their dependencies and read their help text (menuconfig/extension).

There is no distinction or taxonomy between Kconfig options that are applicable to regular |NCS| only and those that are only applicable to |BMlong|.
Kconfig options that are applicable to the current application are shown and are selectable, while others are not.

There is no consistent mapping of the :file:`sdk_config.h` entries to Kconfig options.
Some libraries that were ported from nRF5 have similar Kconfig options as the :file:`sdk_config.h` entries they had in nRF5, but it is not a consistent rule.

In nRF5, the same application/sample had a project file for each supported board and IDE/compiler.
In |BMshort|, there is a single project file (consisting of :file:`CMakeLists.txt` and :file:`prj.conf`) that can be built for different boards with a different command-line instruction or by selecting a different board target in the VS Code extension.

> If necessary, Kconfig options can be specified in a different :file:`.conf` file that is then automatically appended to the default :file:`prj.conf` file, thus realizing a dedicated configuration for a specific board.

Reviewer: Maybe we can mention here that these are called `Kconfig fragments`. We are already referring to these in the main documentation.

Once the application is built, flashing of the SoftDevice must be performed manually.
This step must be performed only once and then the application can be flashed multiple times without having to reflash the SoftDevice.

Memory partitioning
===================

In nRF5, memory partitioning was done using linker scripts.
In |BMshort|, there are a few ready-made partitioning schemes that can be selected by compiling for specific board targets that cover the most common use cases.
Partitioning can be tweaked by making simple changes to textual files called **Devicetree**, which define the layout of the memory.
These can be edited in the board files, or applied to existing boards as **overlays**.

Libraries and drivers
*********************

Whereas nRF5 supported different short range protocols such as Gazell, ESB, Ant, and NFC, those are not supported by |BMshort|.
In general, |BMshort| support focuses on Bluetooth Low Energy.

Bluetooth LE libraries
======================

Bluetooth LE features that are natively offered by the SoftDevice are mostly unchanged from the nRF5, and the SoftDevice documentation highlights any differences.

> As for the collection of Bluetooth Low Energy libraries that were available in the nRF5, |BMshort| offers a limited subset, where each service may have slightly different API and functionality compared to their respective nRF5 implementation.

Reviewer: Nit
Nit:
```suggestion
As for the collection of Bluetooth LE libraries that were available in the nRF5, |BMshort| offers a limited subset, where each service may have slightly different API and functionality compared to their respective nRF5 implementation.
```

The Bluetooth LE services offered in |BMshort| are the following:

* Heart Rate Monitor (peripheral)
* Nordic UART (NUS) (peripheral)
* Nordic LED Button (LBS) (peripheral)
* Continuous Glucose Monitor (peripheral)
* Bond Management
* MCUMgr service (DFU)

> * Battery (peripheral)

Author: Changed to this  , not sure how much more we can group those:

* Peripheral services:

  * Heart Rate Monitor (peripheral)
  * Nordic UART (NUS) (peripheral)
  * Nordic LED Button (LBS) (peripheral)
  * Continuous Glucose Monitor (peripheral)
  * Battery (peripheral)

* MCUMgr service (DFU service)
* Bond Management
* Device Information

.. _nrf5_bm_migration:

Migration notes - nRF5 SDK to nRF Connect SDK Bare Metal
########################################################

> This document outlines the high-level differences between nRF5 SDK and the |BMlong|.

Author: Adding:

.. note::

   This document is in development and being constantly updated.

It is meant to provide support when migrating an application built on nRF5 SDK to |BMshort|.

Project files
*************

In nRF5, project files are tightly coupled with the SDK directory structure.

Typically, projects are placed inside the :file:`examples` folder within the SDK and include header files, source files, and pre-built libraries in defined locations within the SDK.
These projects rely on embedded IDEs project files, such Keil µVision or IAR, or GNU Makefiles, which all have very limited dependency management.
Developers have to manually handle their project dependencies, adding libraries and their relative headers to their project by navigating the IDE’s menu, which can be tedious for large projects.

sdk_config.h to Kconfig
=======================

Project configuration is achieved through the :file:`sdk_config.h` file, which sets the configuration options present in the nRF5 SDK that are applicable to a given project.

> When new dependencies are manually added to the project, their relative configuration entries in the project’s initial :file:`sdk_config.h` file are missing, and must also be added manually.

Author: true, I removed the headings for now, I think they only created confusion

That is done using the :file:`prj.conf` file that replaces the :file:`sdk_config.h` file from the nRF5 SDK.
The entries in the :file:`prj.conf` file are referred to as **Kconfig options**.

Unlike the :file:`sdk_config.h` file that lists all the configuration options relevant for an application, even those whose values are unchanged from defaults, the :file:`prj.conf` file only contains entries whose values are manually set or overridden.

The configuration options whose values are left as default are not present in that file, although when a project is built, a file containing all configuration options pertaining to the application (called :file:`autoconf.h`) is created in the background.
The build system and the |nRFVSC| extension both provide a way to conveniently browse and search all available project options and inspect their dependencies and read their help text (menuconfig/extension).

There is no distinction or taxonomy between Kconfig options that are applicable to regular |NCS| only and those that are only applicable to |BMlong|.

> Kconfig options that are applicable to the current application are shown and are selectable, while others are not.

Reviewer 2: Note/warning?

Author: I would leave as is, let's go back to this later.

In nRF5, the same application/sample had a project file for each supported board and IDE/compiler.
In |BMshort|, there is a single project file (consisting of :file:`CMakeLists.txt` and :file:`prj.conf`) that can be built for different boards with a different command-line instruction or by selecting a different board target in the VS Code extension.
If necessary, Kconfig options can be specified in a different :file:`.conf` file that is then automatically appended to the default :file:`prj.conf` file, thus realizing a dedicated configuration for a specific board.
Kconfig options appended in this way are referred to as **Kconfig fragments**.

Once the application is built, flashing of the SoftDevice must be performed manually.
This step must be performed only once and then the application can be flashed multiple times without having to reflash the SoftDevice.

Memory partitioning
===================

In nRF5, memory partitioning was done using linker scripts.
In |BMshort|, there are a few ready-made partitioning schemes that can be selected by compiling for specific board targets that cover the most common use cases.

> Partitioning can be tweaked by making simple changes to textual files called **Devicetree**, which define the layout of the memory.

Reviewer 2: Files are not called Devicetree, they are Devicetree files.

These can be edited in the board files, or applied to existing boards as **overlays**.

Libraries and drivers
*********************

Whereas nRF5 supported different short range protocols such as Gazell, ESB, Ant, and NFC, those are not supported by |BMshort|.
In general, |BMshort| support focuses on Bluetooth Low Energy.

Bluetooth LE libraries
======================

Bluetooth LE features that are natively offered by the SoftDevice are mostly unchanged from the nRF5, and the SoftDevice documentation highlights any differences.
As for the collection of Bluetooth LE libraries that were available in the nRF5, |BMshort| offers a limited subset, where each service may have slightly different API and functionality compared to their respective nRF5 implementation.

> The Bluetooth LE services offered in |BMshort| are the following:

Reviewer 2: ```suggestion
The Bluetooth LE services offered in |BMshort| at the moment are the following:
```

* Peripheral services:

  * Heart Rate Monitor (peripheral)
  * Nordic UART (NUS) (peripheral)
  * Nordic LED Button (LBS) (peripheral)
  * Continuous Glucose Monitor (peripheral)
  * Battery (peripheral)

* MCUMgr service (DFU service)
* Bond Management
* Device Information

> Utility libraries for Bluetooth LE are available in |BMshort|, but again, their collection may not be as complete, and their functionality and API may be slightly different than their respective nRF5 implementation.

Reviewer 2: This is too informal in my opinion. 
```suggestion
Utility libraries for Bluetooth LE are available in |BMshort|, though their collection may not be as complete, and their functionality and API may be slightly different than their respective nRF5 implementation.
```

> See table below for a summary of supported libraries.

Reviewer 2: This list does not cover all libraries from nRF5, where some that we state no will come in the future. Should this be rewritten in a way that makes that clear? Or can we remove those that are not supported at the moment?

.. list-table:: Supported libraries
   :header-rows: 1

   * - Name
     - Supported
     - New name
   * - ``ble_advertising``
     - Yes
     - ``ble_adv``
   * - ``ble_advdata``
     - Yes
     - Merged with ``ble_adv``
   * - ``ble_db_discovery``
     - No
     - N/A
   * - ``ble_conn_params``
     - Yes
     - Name unchanged
   * - ``ble_conn_state``
     - Yes
     - Name unchanged
   * - ``ble_dtm``
     - No
     - N/A
   * - ``ble_racp``
     - Yes
     - Name unchanged
   * - ``ble_srv_common``
     - No
     - N/A
   * - ``nrf_ble_gatt``
     - Yes
     - ``ble_conn_params``
   * - ``nrf_ble_gq``
     - Yes
     - ``ble_gq``
   * - ``nrf_ble_qwr``
     - Yes
     - ``ble_qwr``
   * - ``nrf_ble_scan``
     - No
     - N/A
   * - ``ble_link_ctx_manager``
     - No
     - N/A
   * - ``ble_radio_notification``
     - No
     - N/A
   * - ``peer_manager``

>      - Yes

Reviewer 2: Not yet!
```suggestion
     - No
```
